### PR TITLE
Bug 1823462 - New application-services release system

### PR DESCRIPTION
**Pull request by bendk**
Application services is going to switch to rapid-release / train style model where:
  - The version numbers correspond to the Fx version (113.x for the current nightly cycle).
  - There are nightly builds for the current nightly release (for example 113.[timestamp]
  - At the end of the nightly release cycle, there's a release (for example 113.0)
  - There may be subsequent releases if changes are uplifted (for example 113.1)

This requires several updates:
- Updated the gradle code to handle the new system.  The application services versioning is stored in the version/channel fields of the `ApplicationServicesConfig` object.
- Added nightly-bump script that reads the latest nightly data from the taskcluster index and automatically updates the `ApplicationServices.kt`.  The plan is to run this manually for a few days to make sure it works, then create a system that automatically runs it and creates a PR, then finally to auto-merge the PR if the CI is green.
- Updated the gradle code to fetch packages from the Maven nightly/staging repos if needed.
- Updated creating-a-release-branch docs

Don't land this one until we agree on the correct process for this and [application-services merges its corresponding PR](https://github.com/mozilla/application-services/pull/5408#pullrequestreview-1324315808) and starts producing nightlies.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1823462

### CODE REVIEW

**Comment by bendk** [comment]
I'm not sure who should review this one.  The one person I was pretty sure of is @rvandermeulen, please pull in any others that you think should be in the discussion.

File: android-components/build.gradle
Comment (bendk) [inline android-components/build.gradle:86]: Nightlies should probably live in a separate location from the other maven packages.  The current proposal is to:
  - Use the `nightly` maven repository for nightly releases
  - Use the `release` maven repository for the actual releases
  - Use the `nightly-staging` and `staging` repositories for preview releases built from PRs.
  - Use groupId  `org.mozilla.appservices` for actual releases
  - Use groupId `org.mozilla.appservices.nightly` for nightly/preview releases.

However, this is very flexible and we can follow whatever the Android team thinks is best.

File: fenix/docs/Creating-a-release-branch.md
Comment (bendk) [inline fenix/docs/Creating-a-release-branch.md:24]: I think this step makes sense, but we can be flexible here as well.  Whatever works for you is good with us.

I think that the Fenix beta release will need to depend on the app-services release.  We can coordinate and make sure our release goes out before you start your release process (maybe the work day before?).

File: android-components/automation/application-services-nightly-bump.py
Comment (bendk) [inline android-components/automation/application-services-nightly-bump.py:58]: As mentioned in the commit message, once we've agreed on everything and the app-services nightlies are running, we should:
- Run this script manually
- Once we figure out any snags to running manually, setup a cron job to run this script, create a commit and PR for it.
- Once that cron job is running smoothly, setup automerge on the PRs.

How does that sound to others?

File: android-components/plugins/dependencies/src/main/java/ExtraRepositories.kt
Comment (csadilek) [inline android-components/plugins/dependencies/src/main/java/ExtraRepositories.kt:12]: What's interesting here is that GeckoView (Nightly) isn't published to `nightly.mozilla.maven.org`. Distinct artifacts are published for nightly|beta|release: 
https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/plugins/dependencies/src/main/java/Gecko.kt#L26

See: https://maven.mozilla.org/?prefix=maven2/org/mozilla/geckoview/

https://nightly.maven.mozilla.org/ has actually only ever been used by A-C. Do you have a preference here? It seems the former would make this a bit easier, as we could use the same approach as Gecko.kt. Not sure though, wdyt?

Comment (csadilek) [reply android-components/plugins/dependencies/src/main/java/ExtraRepositories.kt:12]: OK, I don't see an advantage in using a separate Maven repo for Nightly. So following the same approach we use for GeckoView (Gecko.kt) makes sense to me, and should simplify this for you.

**Comment by bendk** [comment]
Updated the code to always use either the production or staging maven repo, let's not try to also use the nightly/nightly-staging repos.

Fixed an issue with the glean version and the tests are passing again.

How is the rest of this looking?  Any concerns with the proposed release process?  I'm want to switch to the new process sometime next week, is there anything we need to figure out before then?

File: fenix/docs/Creating-a-release-branch.md
Comment (rvandermeulen) [reply fenix/docs/Creating-a-release-branch.md:24]: https://github.com/mozilla-mobile/firefox-android/blob/main/docs/contribute/release_checklist.md are the docs we should updating here.

Comment (rvandermeulen) [reply fenix/docs/Creating-a-release-branch.md:24]: Right now, our practice for GeckoView is to do the s/NIGHTLY/BETA change immediately after creating the branch on merge day, then let relbot do the actual version bump once the first beta build is uploaded to Maven later on. I'm not sure if this approach is feasible here, though, since we'd be switching from XXX.YYYYMMDD format to XXX.0 also, but maybe?

But regardless, I think the right answer here with respect to timing is that the A-S team would do their branching and release tagging on merge day alongside the Gecko & firefox-android branching and then we can do the firefox-android bump afterwards.

**Review by rvandermeulen — CHANGES_REQUESTED** [review]
Left a couple comments that I think should be addressed, but I also want to be clear that my review here is strictly about process and not on the specific code itself.

File: android-components/automation/application-services-nightly-bump.py
Comment (bendk) [reply android-components/automation/application-services-nightly-bump.py:58]: That sounds great to me.  So would this mean updating the `android_components._update_application_services()` function to pull in the latest nightly using the same basic process as the `application-services-nightly-bump.py` script?

Comment (rvandermeulen) [reply android-components/automation/application-services-nightly-bump.py:58]: In a perfect world, it'll also be able to handle the Nightly major version change too once that bump hits Maven.

Comment (bendk) [reply android-components/automation/application-services-nightly-bump.py]: Relbot definitely seems like the right place for this, I made this PR in that repo: https://github.com/mozilla-mobile/relbot/pull/105

File: fenix/docs/Creating-a-release-branch.md
Comment (ddurst) [reply fenix/docs/Creating-a-release-branch.md]: @rvandermeulen Looks like there's only this one question about what happens now with GV. ^ Seems like the last detail to address.

Comment (rvandermeulen) [reply fenix/docs/Creating-a-release-branch.md]: For GeckoView, the Nightly builds upload to a different directory than the Beta & Release ones, so they don't end up living side-by-side in the same place.
https://maven.mozilla.org/?prefix=maven2/org/mozilla/geckoview/

File: android-components/plugins/dependencies/src/main/java/ApplicationServices.kt
Comment (csadilek) [inline android-components/plugins/dependencies/src/main/java/ApplicationServices.kt:18]: nit: We could remove the comment about the workaround, because it's fine to have a separate groupId. That's what we do in GV as well. :)

**Review by csadilek — COMMENTED** [review]
> @csadilek is there any other issues to address before this gets merged? I really hope to get this merged tomorrow and start testing the nightlies on Monday.

@bendk Thank you for working through this! Let's not land this on a Friday though :). From my perspective, it's OK to land this on Monday and test it out. So, you're not losing any time here. 

We also want @rvandermeulen to take another look :) and approve before landing. Thank you!

**Comment by bendk** [comment]
> Thank you for working through this! Let's not land this on a Friday though :). From my perspective, it's OK to land this on Monday and test it out. So, you're not losing any time here.

Also a good point.  I didn't want to wait for the cron schedule to trigger, but relbot helps out here.  I'll aim to merge this on Monday, then I'll test if the PR looks correct using relbot + my personal firefox-android fork.

**Comment by bendk** [comment]
For reference, here's how the new versions will display in the about page:
image

.. and here's how the maven repo will look (except it will use the production maven repo once the app-services PR lands).
https://maven-default.stage.mozaws.net/?prefix=maven2/org/mozilla/appservices/nightly/full-megazord/

**Comment by bendk** [comment]
@rvandermeulen and I discussed adding a `-0a1` to the nightly about page.  I'm very open to that as a follow-up, but I'd like to avoid making any more changes to this PR.

**Review by rvandermeulen — APPROVED** [review]
Yeah, I think showing a display version that's consistent with GV would be good, but I don't want to block this PR landing over it. r=me, reiterating again that my review here is strictly about process/end-result and not on the specific code itself!

**Review by csadilek — APPROVED** [review]
Forgot to ✅ this. Let's land and try out next week, as discussed.

**Comment by bendk** [comment]
Is this ready to land now?  I updated the PR to use the latest nightly.  I'm not sure why the tests are failing, it seems unlikely to be related to the changes here.

**Pull request merged**
